Maple2.Server.Game/PacketHandlers/UgcHandler.cs

```diff
@@ -61,25 +67,97 @@ private enum Command : byte {
                 for (int i = 0; i < hours; i++) {
                     var reservation = packet.Read<UgcBannerReservation>();
                 }
-                break;
+                return;
+            default:
+                Logger.Information("Unimplemented Ugc Type: {type}", info.Type);
+                return;
+        }
+    }
+
+    private void UploadItem(GameSession session, IByteReader packet, UgcType ugcType) {
+        long itemUid = packet.ReadLong();
+        int itemId = packet.ReadInt();
+        int amount = packet.ReadInt();
+        string name = packet.ReadUnicodeString();
+        packet.ReadByte();
+        long cost = packet.ReadLong();
+        bool useVoucher = packet.ReadBool();
+
+        if (!TableMetadata.UgcDesignTable.Entries.TryGetValue(itemId, out UgcDesignTable.Entry? ugcMetadata)) {
+            return;
+        }
+
+        if (useVoucher) {
+            Item? voucher = session.Item.Inventory.Filter(item => item.Metadata.Property.Tag == ItemTag.FreeDesignCoupon).FirstOrDefault();
+            if (voucher != null) {
+                if (!session.Item.Inventory.Consume(voucher.Uid, 1)) {
+                    session.Send(NoticePacket.Notice(NoticePacket.Flags.Alert | NoticePacket.Flags.Message, StringCode.s_err_invalid_item));
+                    return;
+                }
+            }
+        } else {
+            switch (ugcMetadata.CurrencyType) {
+                case MeretMarketCurrencyType.Meso:
+                    if (session.Currency.CanAddMeso(-ugcMetadata.CreatePrice) != -ugcMetadata.CreatePrice) {
+                        session.Send(NoticePacket.MessageBox(StringCode.s_err_lack_meso));
+                        return;
+                    }
+                    session.Currency.Meso -= ugcMetadata.CreatePrice;
+                    break;
+                case MeretMarketCurrencyType.Meret:
+                    if (session.Currency.CanAddMeret(-ugcMetadata.CreatePrice) != -ugcMetadata.CreatePrice) {
+                        session.Send(NoticePacket.MessageBox(StringCode.s_err_lack_merat));
+                        return;
+                    }
+                    session.Currency.Meret -= ugcMetadata.CreatePrice;
+                    break;
+                case MeretMarketCurrencyType.RedMeret:
+                    if (session.Currency.CanAddGameMeret(-ugcMetadata.CreatePrice) != -ugcMetadata.CreatePrice) {
+                        session.Send(NoticePacket.MessageBox(StringCode.s_err_lack_merat_red));
+                    }
+                    session.Currency.GameMeret -= ugcMetadata.CreatePrice;
+                    break;
+                default:
+                    Logger.Error("Unhandled currency type {UgcMetadataCurrencyType}", ugcMetadata.CurrencyType);
+                    return;
+            }
+        }
+
+        Item? item = session.Item.CreateItem(itemId, ugcMetadata.ItemRarity);
+        if (item == null) {
+            return;
```

Comment:
This needs to be logged as fatal since the player already paid for this we shouldn't silently fail.

something similar to:
```cs
Log.Fatal("Failed to consume ingredient {ItemUid}", ingredient.Uid);
throw new InvalidOperationException($"Fatal: Consuming ingredient: {ingredient.Uid}");
```